src/Membership.sol

```diff
@@ -0,0 +1,427 @@
+
+    /// @notice Emitted when a membership in grace period is extended
+    /// @param idCommitment the idCommitment of the member
+    /// @param userMessageLimit the rate limit of this membership
+    /// @param index the index of the membership in the merkle tree
+    /// @param newExpirationDate the new expiration date of this membership
+    event MemberExtended(uint256 idCommitment, uint32 userMessageLimit, uint32 index, uint256 newExpirationDate);
+
+    /// @dev contract initializer
+    /// @param _priceCalculator Address of an instance of IPriceCalculator
+    /// @param _maxTotalRateLimitPerEpoch Maximum total rate limit of all memberships in the tree
+    /// @param _minRateLimitPerMembership Minimum rate limit of one membership
+    /// @param _maxRateLimitPerMembership Maximum rate limit of one membership
+    /// @param _expirationTerm Membership expiration term
+    /// @param _gracePeriod Membership grace period
+    function __Membership_init(
+        address _priceCalculator,
+        uint32 _maxTotalRateLimitPerEpoch,
+        uint32 _minRateLimitPerMembership,
+        uint32 _maxRateLimitPerMembership,
+        uint32 _expirationTerm,
+        uint32 _gracePeriod
+    )
+        internal
+    {
+        priceCalculator = IPriceCalculator(_priceCalculator);
+        maxTotalRateLimitPerEpoch = _maxTotalRateLimitPerEpoch;
+        maxRateLimitPerMembership = _maxRateLimitPerMembership;
+        minRateLimitPerMembership = _minRateLimitPerMembership;
+        billingPeriod = _expirationTerm;
+        gracePeriod = _gracePeriod;
+    }
+
+    /// @notice Checks if a user message limit is valid. This does not take into account whether we the total
+    /// memberships have reached already the `maxTotalRateLimitPerEpoch`
+    /// @param userMessageLimit The user message limit
+    /// @return true if the user message limit is valid, false otherwise
+    function isValidUserMessageLimit(uint32 userMessageLimit) external view returns (bool) {
+        return userMessageLimit >= minRateLimitPerMembership && userMessageLimit <= maxRateLimitPerMembership;
+    }
+
+    /// @dev acquire a membership and trasnfer the fees to the contract
+    /// @param _sender address of the owner of the new membership
+    /// @param _idCommitment the idcommitment of the new membership
+    /// @param _rateLimit the user message limit
+    /// @param _numberOfPeriods the number of periods the user wants to acquire
+    /// @return index the index in the merkle tree
+    /// @return reusedIndex indicates whether a new leaf is being used or if using an existing leaf in the merkle tree
+    function _acquireMembership(
+        address _sender,
+        uint256 _idCommitment,
+        uint32 _rateLimit,
+        uint32 _numberOfPeriods
+    )
+        internal
+        returns (uint32 index, bool reusedIndex)
+    {
+        if (_numberOfPeriods == 0) revert NumberOfPeriodsCantBeZero();
```

Comment:
Should we set a maximum too?

Reply:
Not sure I totally understand what you mean by "period" but when acquiring a membership, the only expiration time available should be 180 days. This value should be a parameter in the smart contract (`Membership expiration term`) that can be changed.

Or does period mean something else?

Reply:
@fryorcraken It's related to this https://github.com/waku-org/specs/pull/34/files#r1743943629

In short, instead of allowing to register a fixed amount of 180 days I'm suggesting to allow registering `x` times a `period`. We can lower this period to eg 30days so that its closer to web2. And this would allow to register for 30*1days, 30*2 days, 30*3days, etc. the user can choose.

This also lowers the entry price. Eg would you pay for a service that requires upfront payment of 180 days? 

This also related to an idea that is being cooked regarding incentives, where the fees of a given "period" are distributed to node operators. Will share once ready.

Reply:
As decided on the research call today (2024-09-09), let's set aside the billing period idea for now, and review it more carefully for the next spec iteration. I've created an [issue](https://github.com/waku-org/specs/issues/37) for that purpose, we can move the discussion there.

Reply:
Billing period idea removed from code in https://github.com/waku-org/waku-rlnv2-contract/pull/13/commits/0cd80d639c1f6b3b0b8311dc6f1c1777e9a36323